ros2topic/ros2topic/verb/hz.py

```diff
@@ -0,0 +1,337 @@
+            return self.msg_t0
+        return self._msg_t0[topic]
+
+    def set_msg_t0(self, value, topic=None):
+        if topic is None:
+            self.msg_t0 = value
+        self._msg_t0[topic] = value
+
+    def get_msg_tn(self, topic=None):
+        if topic is None:
+            return self.msg_tn
+        return self._msg_tn[topic]
+
+    def set_msg_tn(self, value, topic=None):
+        if topic is None:
+            self.msg_tn = value
+        self._msg_tn[topic] = value
+
+    def get_times(self, topic=None):
+        if topic is None:
+            return self.times
+        return self._times[topic]
+
+    def set_times(self, value, topic=None):
+        if topic is None:
+            self.times = value
+        self._times[topic] = value
+
+    def callback_hz(self, m, topic=None):
+        """
+        Calculate interval time.
+
+        :param m: Message instance
+        :param topic: Topic name
+        """
+        # ignore messages that don't match filter
+        if self.filter_expr is not None and not self.filter_expr(m):
+            return
+        with self.lock:
+            # Uses ROS time as the default time source and Walltime only if requested
+            curr_rostime = self._clock.now() if not self.use_wtime else \
+                Clock(clock_type=ClockType.SYSTEM_TIME).now()
+
+            # time reset
+            if curr_rostime.nanoseconds == 0:
+                if len(self.get_times(topic=topic)) > 0:
+                    print('time has reset, resetting counters')
+                    self.set_times([], topic=topic)
+                return
+
+            curr = curr_rostime.nanoseconds
+            if self.get_msg_t0(topic=topic) < 0 or self.get_msg_t0(topic=topic) > curr:
+                self.set_msg_t0(curr, topic=topic)
+                self.set_msg_tn(curr, topic=topic)
+                self.set_times([], topic=topic)
+            else:
+                self.get_times(topic=topic).append(curr - self.get_msg_tn(topic=topic))
+                self.set_msg_tn(curr, topic=topic)
+
+            if len(self.get_times(topic=topic)) > self.window_size:
+                self.get_times(topic=topic).pop(0)
+
+    def get_hz(self, topic=None):
+        """
+        Calculate the average publising rate.
+
+        :param topic: topic name, ``list`` of ``str``
+        :returns: tuple of stat results
+            (rate, min_delta, max_delta, standard deviation, window number)
+            None when waiting for the first message or there is no new one
+        """
+        if not self.get_times(topic=topic):
+            return
+        elif self.get_last_printed_tn(topic=topic) == 0:
+            self.set_last_printed_tn(self.get_msg_tn(topic=topic), topic=topic)
+            return
+        elif self.get_msg_tn(topic=topic) < self.get_last_printed_tn(topic=topic) + 1e9:
+            return
+        with self.lock:
+            # Get frequency every one minute
+            n = len(self.get_times(topic=topic))
```

Comment:
Again, probably more efficient to call `self.get_times(topic=topic)` right after taking the lock, then reuse the local variable.

Reply:
fixed in 0ba39d3